fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBDatabaseRunner.java

```diff
@@ -537,44 +284,51 @@ public T run(@Nonnull Function<? super FDBRecordContext, ? extends T> retriable)
-            final Exception exception = new RunnerClosed();
-            for (CompletableFuture<?> future : futuresToCompleteExceptionally) {
-                future.completeExceptionally(exception);
-            }
-        }
-        contextsToClose.forEach(FDBRecordContext::close);
-    }
+    void close();
 
-    private synchronized void addContextToClose(@Nonnull FDBRecordContext context) {
-        if (closed) {
-            context.close();
-            throw new RunnerClosed();
-        }
-        contextsToClose.removeIf(FDBRecordContext::isClosed);
-        contextsToClose.add(context);
-    }
+    /**
+     * <p>
+     * Produces a new runner, wrapping this runner, which performs all work in the context of a new
+     * {@link SynchronizedSession}.
+     * </p>
+     * <p>
+     * The returned runner will have acquired and started the lease, so care must be taken to ensure that
+     * work begins before the lease expiration period.
+     * </p>
+     * @param lockSubspace the lock for which the session contends
+     * @param leaseLengthMillis length between last access and lease's end time in milliseconds
+     * @return a future that will return a runner maintaining a new synchronized session
+     * @see SynchronizedSession
+     */
+    @API(API.Status.EXPERIMENTAL)
+    CompletableFuture<SynchronizedSessionRunner> startSynchronizedSessionAsync(@Nonnull Subspace lockSubspace, long leaseLengthMillis);
 
-    private synchronized void addFutureToCompleteExceptionally(@Nonnull CompletableFuture<?> future) {
-        if (closed) {
-            final RunnerClosed exception = new RunnerClosed();
-            future.completeExceptionally(exception);
-            throw exception;
-        }
-        futuresToCompleteExceptionally.removeIf(CompletableFuture::isDone);
-        futuresToCompleteExceptionally.add(future);
-    }
+    /**
+     * Synchronous/blocking version of {@link #startSynchronizedSessionAsync(Subspace, long)}.
+     * @param lockSubspace the lock for which the session contends
+     * @param leaseLengthMillis length between last access and lease's end time in milliseconds
+     * @return a runner maintaining a new synchronized session
+     */
+    @API(API.Status.EXPERIMENTAL)
+    SynchronizedSessionRunner startSynchronizedSession(@Nonnull Subspace lockSubspace, long leaseLengthMillis);
+
+    /**
+     * Produces a new runner, wrapping this runner, which performs all work in the context of an existing
+     * {@link SynchronizedSession}.
+     * @param lockSubspace the lock for which the session contends
+     * @param sessionId session ID
+     * @param leaseLengthMillis length between last access and lease's end time in milliseconds
+     * @return a runner maintaining a existing synchronized session
+     * @see SynchronizedSession
+     */
+    @API(API.Status.EXPERIMENTAL)
+    SynchronizedSessionRunner joinSynchronizedSession(@Nonnull Subspace lockSubspace, @Nonnull UUID sessionId, long leaseLengthMillis);
```

Comment:
I suppose you could also imagine having `joinSyncrhonizedSession(@Nonnull SyncrhonizedSession session)` variants of this method. I suppose it's not necessary. I'm not sure which one would make it easier down the line, and it's obviously something that could be added later.

Reply:
Ah, sorry I misread the parameter. Yes, I agree.